<div style="border:solid black 2px; padding: 20px"><h1 style="color:green; margin-bottom:20px">Reviewer's comment v1</h1>
Halo Chulsum!

Perkenalkan saya Naufal, disini saya akan mereview tugas Chulsum ya.

Saya akan memberikan beberapa komen seperti dibawah ini, tolong jangan dipindah, dirubah, maupun dihapus ya.

Komen yang saya berikan akan muncul dalam warna hijau, kuning, atau pun merah seperti ini:

<div class="alert alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

<div class="alert alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Terdapat beberapa catatan.

</div>

<div class="alert alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Perlu beberapa perbaikan

</div>

Dan tentu saja Chulsum dapat menjawab saya dengan menggunakan ini

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
</div>

<div style="border:solid black 2px; padding: 20px">
<b>Reviewer's comment v1:</b>
    
<b>Overall Feedback</b> 
    
- Project ini merupakan lompatan yang sangat jauh dibandingkan project sebelumnya maka dari itu saya berterima kasih telah memberikan waktu dan usaha yang tidak sedikit dalam pengerjaannya.
- Namun terdapat beberapa poin kritikal seperti intruksi-instruksi yang ada sebaiknya ditulis ulang sedemikian rupa agar seolah-olah Chulsum sendiri yang memiliki inisiatif dalam alur pengerjaan project ini dan melatih sense of business dalam konteks project ini. 
    
Untuk perbaikan nya ditunggu ya!
</div>

        

<div style="border:solid black 2px; padding: 20px">
<b>Reviewer's comment v2:</b>
    
<b>Overall Feedback</b> 
    
Well-done! Perbaikan yang dilakukan sudah bagus dan efisien. Dengan begitu project ini bisa disetujui dan melanjutkan ke tahapan berikutnya. Walaupun masih terdapat catatan bahwa fungsi yang dibuat masih belum sepenuhnya efisien namun bisa menjadi bahan untuk improvement kedepanya. Keep up the good work ya! 💪💪
</div>

        

# Menganalisis risiko peminjam gagal membayar

Proyek Anda ialah menyiapkan laporan untuk bank bagian kredit. Anda harus mencari tahu pengaruh status perkawinan seorang nasabah dan jumlah anak terhadap probabilitas ketepatan waktu dalam melunasi pinjaman. Bank sudah memiliki beberapa data mengenai kelayakan kredit nasabah.

Laporan Anda akan dipertimbangkan pada saat membuat **penilaian kredit** untuk calon nasabah. **Penilaian kredit** digunakan untuk mengevaluasi kemampuan calon peminjam untuk melunasi pinjaman mereka.

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Sebelum melangkah lebih jauh dalam project ini, instruksi atau petunjuk yang ditandai dalam `[]` bisa dihapus ya. Sebaiknya Chulsum menjelaskan terkait projek ini terkait tujuan dan langkah-langkah yang akan diambil dalam pengerjaanya.

</div>

## Buka *file* data dan baca informasi umumnya. 


In [1]:
# memuat library

import pandas as pd, numpy as np

In [2]:
# memuat data

path = '/datasets/credit_scoring_eng.csv'
df = pd.read_csv(path)

df

children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  Secondary Education             1   
3             3   -4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  
0         40620.102    purchase of the house  
1         17932.802             car purchase  
2         23341.752    purchase of the house  
3         42820.568  supplementary education  
4         25378.572        to have a wedding  
...             ...                      ...  
21520     35966.698     housing transactions  
21521     24959.969        purchase of a car  
21522     14347.610                 property  
21523     39054.888        buying my own car  
21524     13127.587             to buy a car  

[21525 rows x 12 columns]

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Akan lebih baik lagi apabila langkah memuat library dan data dipisah di dua sel yang berbeda. Hal ini akan meminimalisir perulangan ketika dilakukan muat ulang data saja tanpa muat ulang library yang sudah termuat sebelumnya.

</div>

## Soal 1. Eksplorasi Data

**Deskripsi Data**
- *`children`* - jumlah anak dalam keluarga
- *`days_employed`* - pengalaman kerja dalam hari
- *`dob_years`* - usia klien dalam tahun
- *`education`* - pendidikan klien
- *`education_id`* - tanda pengenal pendidikan
- *`family_status`* - status perkawinan
- *`family_status_id`* - tanda pengenal status perkawinan
- *`gender`* - jenis kelamin klien
- *`income_type`* - jenis pekerjaan
- *`debt`* - apakah klien memiliki hutang pembayaran pinjaman
- *`total_income`* - pendapatan bulanan
- *`purpose`* - tujuan mendapatkan pinjaman

In [3]:
# banyak baris dan kolom yang dimiliki oleh dataset

df.shape

(21525, 12)

In [4]:
# menampilkan 5 baris pertama

df.head(5)

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

Terdapat beberapa nilai pada data yang bermasalah. Salah satunya yaitu pada kolom `days_employed` terdapat nilai negatif padahal deskripsi dari kolom tersebut adalah pengalaman kerja dalam hari (tidak mungkin hari negatif). Bisa jadi terdapat data bermasalah lain pada data, maka dari itu diperlukan pengecekan dan pembetulan lebih lanjut

In [5]:
# mendapatkan informasi data

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [6]:
# menampilkan jumlah nilai kosong tiap kolom 

df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Hindari menampilkan multipel output pada satu sel. Pemisahan satu output dalam satu sel akan mempermudah memahami konteks di setiap output dan dapat dijelaskan step by step nya.
</div>

Terdapat nilai hilang di kolom `days_employed` dan `total_income` dengan jumlah yang sama

In [7]:
# menampilkan tabel yang telah difilter dengan nilai yang hilang di kolom days_employed dengan data yang hilang

df.loc[df['days_employed'].isna()] 

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

Setelah dilakukan pemfilteran pada baris dengan nilai kosong diketahui bahwa letak nilai kosong pada tiap baris sama-sama berada di kedua kolom tersebut (simetris). Untuk memastikan asumsi tersebut, perlu dilakukan pengecekan lebih lanjut

In [8]:
# menerapkan beberapa persyaratan untuk memfilter data dan melihat jumlah baris dalam tabel yang difilter.

data_null = df.loc[(df['days_employed'].isna()) & (df['total_income'].isna())]
data_null

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

**Kesimpulan sementara** 

Jumlah baris dalam tabel yang difilter sesuai dengan jumlah nilai yang hilang. Ada kemungkinan kolom `days_employed` dan `total_income` saling berhubungan tetapi masih belum dapat dipastikan apa hubungan dari kedua kolom tersebut.

Langkah selanjutnya yaitu memeriksa persentase nilai yang hilang dengan seluruh kumpulan data, kemudian membandingkan value dan distribusi antara data seluruh dataset dengan data yang terfilter berdasarkan nilai kosong 

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Kemungkinan masih ada error di bagian translasi, sebaiknya menggunakan `kesimpulan sementara` dibandingkan `kesimpulan menengah`.
</div>

In [9]:
# Memeriksa persentase nilai yang hilang yang dibandingkan dengan seluruh kumpulan data

round(df.isna().sum() / len(df) * 100, 2)

children             0.0
days_employed       10.1
dob_years            0.0
education            0.0
education_id         0.0
family_status        0.0
family_status_id     0.0
gender               0.0
income_type          0.0
debt                 0.0
total_income        10.1
purpose              0.0
dtype: float64

In [10]:
# Menampilkan unique value di kolom income_type yang terdapat missing value pada kolom total_income nya

sorted(df.loc[df['total_income'].isna(), 'income_type'].unique())

['business', 'civil servant', 'employee', 'entrepreneur', 'retiree']

In [11]:
# Menampilkan unique value di kolom income_type

sorted(df['income_type'].unique())

['business',
 'civil servant',
 'employee',
 'entrepreneur',
 'paternity / maternity leave',
 'retiree',
 'student',
 'unemployed']

In [12]:
# sample distribusi data dari salah satu value di kolom income_type yang terdapat missing value pada kolom total_income nya

df.loc[(df['income_type']=='retiree') & (df['total_income'].isna())]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
29            0            NaN         63  secondary education             1   
55            0            NaN         54  secondary education             1   
67            0            NaN         52    bachelor's degree             0   
145           0            NaN         62  secondary education             1   
...         ...            ...        ...                  ...           ...   
21311         0            NaN         49  secondary education             1   
21321         0            NaN         56  Secondary Education             1   
21414         0            NaN         65  secondary education             1   
21415         0            NaN         54  secondary education             1   
21423         0            NaN         63  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
12     civil partnership                 1      M     retiree     0   
29             unmarried                 4      F     retiree     0   
55     civil partnership                 1      F     retiree     1   
67               married                 0      F     retiree     0   
145              married                 0      M     retiree     0   
...                  ...               ...    ...         ...   ...   
21311            married                 0      F     retiree     0   
21321            married                 0      F     retiree     0   
21414            married                 0      F     retiree     0   
21415            married                 0      F     retiree     0   
21423            married                 0      M     retiree     0   

       total_income                              purpose  
12              NaN                    to have a wedding  
29              NaN               building a real estate  
55              NaN                    to have a wedding  
67              NaN  purchase of the house for my family  
145             NaN                  building a property  
...             ...                                  ...  
21311           NaN      buying property for renting out  
21321           NaN             real estate transactions  
21414           NaN             purchase of my own house  
21415           NaN                 housing transactions  
21423           NaN                    purchase of a car  

[413 rows x 12 columns]

In [13]:
# Menampilkan unique value di kolom gender yang terdapat missing value pada kolom total_income nya

sorted(df.loc[df['total_income'].isna(), 'gender'].unique())

['F', 'M']

In [14]:
# Menampilkan unique value di kolom gender

sorted(df['gender'].unique())

['F', 'M', 'XNA']

In [15]:
# sample distribusi data dari salah satu value di kolom gender yang terdapat missing value pada kolom total_income nya

df.loc[(df['gender']=='F') & (df['total_income'].isna())]

children  days_employed  dob_years            education  education_id  \
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
67            0            NaN         52    bachelor's degree             0   
82            2            NaN         50    bachelor's degree             0   
...         ...            ...        ...                  ...           ...   
21432         1            NaN         38         some college             2   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
67               married                 0      F        retiree     0   
82               married                 0      F       employee     0   
...                  ...               ...    ...            ...   ...   
21432          unmarried                 4      F       employee     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                              purpose  
29              NaN               building a real estate  
41              NaN             second-hand car purchase  
55              NaN                    to have a wedding  
67              NaN  purchase of the house for my family  
82              NaN                              housing  
...             ...                                  ...  
21432           NaN                 housing transactions  
21495           NaN                     wedding ceremony  
21497           NaN                  building a property  
21502           NaN               building a real estate  
21510           NaN                         car purchase  

[1484 rows x 12 columns]

In [16]:
# Menampilkan unique value di kolom family_status yang terdapat missing value pada kolom total_income nya

sorted(df.loc[df['total_income'].isna(), 'family_status'].unique())

['civil partnership', 'divorced', 'married', 'unmarried', 'widow / widower']

In [17]:
# Menampilkan unique value di kolom family_status

sorted(df['family_status'].unique())

['civil partnership', 'divorced', 'married', 'unmarried', 'widow / widower']

In [18]:
# sample distribusi data dari salah satu value di kolom family_status yang terdapat missing value pada kolom total_income nya

df.loc[(df['family_status']=='civil partnership') & (df['total_income'].isna())]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
55            0            NaN         54  secondary education             1   
94            1            NaN         34    bachelor's degree             0   
141           0            NaN         39  secondary education             1   
181           0            NaN         26  secondary education             1   
...         ...            ...        ...                  ...           ...   
21268         1            NaN         44    bachelor's degree             0   
21271         2            NaN         42  secondary education             1   
21300         2            NaN         45  Secondary Education             1   
21463         1            NaN         35    bachelor's degree             0   
21495         1            NaN         50  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
55     civil partnership                 1      F        retiree     1   
94     civil partnership                 1      F       business     0   
141    civil partnership                 1      M       employee     0   
181    civil partnership                 1      F       business     1   
...                  ...               ...    ...            ...   ...   
21268  civil partnership                 1      F  civil servant     0   
21271  civil partnership                 1      M       employee     1   
21300  civil partnership                 1      M       business     0   
21463  civil partnership                 1      M       employee     0   
21495  civil partnership                 1      F       employee     0   

       total_income                              purpose  
12              NaN                    to have a wedding  
55              NaN                    to have a wedding  
94              NaN                     having a wedding  
141             NaN                     wedding ceremony  
181             NaN  purchase of the house for my family  
...             ...                                  ...  
21268           NaN                     having a wedding  
21271           NaN     transactions with my real estate  
21300           NaN     to get a supplementary education  
21463           NaN                     having a wedding  
21495           NaN                     wedding ceremony  

[442 rows x 12 columns]

In [19]:
# Menampilkan unique value di kolom education yang terdapat missing value pada kolom total_income nya

sorted(df.loc[df['total_income'].isna(), 'education'].unique())

["BACHELOR'S DEGREE",
 "Bachelor's Degree",
 'PRIMARY EDUCATION',
 'Primary Education',
 'SECONDARY EDUCATION',
 'SOME COLLEGE',
 'Secondary Education',
 'Some College',
 "bachelor's degree",
 'primary education',
 'secondary education',
 'some college']

In [20]:
# Menampilkan unique value di kolom education

sorted(df['education'].unique())

["BACHELOR'S DEGREE",
 "Bachelor's Degree",
 'GRADUATE DEGREE',
 'Graduate Degree',
 'PRIMARY EDUCATION',
 'Primary Education',
 'SECONDARY EDUCATION',
 'SOME COLLEGE',
 'Secondary Education',
 'Some College',
 "bachelor's degree",
 'graduate degree',
 'primary education',
 'secondary education',
 'some college']

In [21]:
# sample distribusi data dari salah satu value di kolom family_status yang terdapat missing value pada kolom total_income nya

df.loc[(df['education']=='primary education') & (df['total_income'].isna())]

children  days_employed  dob_years          education  education_id  \
1303          1            NaN         70  primary education             3   
1510          0            NaN         52  primary education             3   
1520          0            NaN         53  primary education             3   
4388          0            NaN         39  primary education             3   
4585          1            NaN         58  primary education             3   
7314          1            NaN         45  primary education             3   
8142          0            NaN         64  primary education             3   
10998         0            NaN         34  primary education             3   
11758         0            NaN         58  primary education             3   
13031         0            NaN         37  primary education             3   
13074         3            NaN         59  primary education             3   
13493         0            NaN         63  primary education             3   
15074         0            NaN         62  primary education             3   
17014         0            NaN         39  primary education             3   
17430         0            NaN         60  primary education             3   
17696         1            NaN         45  primary education             3   
18852         0            NaN         26  primary education             3   
20167         0            NaN         36  primary education             3   
20283         0            NaN         61  primary education             3   

           family_status  family_status_id gender    income_type  debt  \
1303   civil partnership                 1      F       employee     0   
1510   civil partnership                 1      M       employee     1   
1520             married                 0      F       employee     0   
4388             married                 0      F       employee     0   
4585   civil partnership                 1      F       employee     0   
7314             married                 0      M       employee     0   
8142   civil partnership                 1      F  civil servant     0   
10998            married                 0      F       employee     0   
11758            married                 0      M        retiree     0   
13031  civil partnership                 1      F       business     1   
13074            married                 0      M       employee     0   
13493            married                 0      F        retiree     0   
15074  civil partnership                 1      F       employee     0   
17014            married                 0      F  civil servant     0   
17430    widow / widower                 2      F        retiree     0   
17696            married                 0      F        retiree     0   
18852          unmarried                 4      M       employee     0   
20167          unmarried                 4      M       employee     0   
20283    widow / widower                 2      F        retiree     0   

       total_income                                   purpose  
1303            NaN  transactions with commercial real estate  
1510            NaN              construction of own property  
1520            NaN          transactions with my real estate  
4388            NaN                       building a property  
4585            NaN                      housing transactions  
7314            NaN                         profile education  
8142            NaN                         to have a wedding  
10998           NaN                           buy real estate  
11758           NaN          transactions with my real estate  
13031           NaN                          wedding ceremony  
13074           NaN                  getting higher education  
13493           NaN                              to own a car  
15074           NaN                  real estate transactions  
17014           NaN  transactions with commercial real estate  
17430

In [22]:
# Menampilkan unique value di kolom education yang terdapat missing value pada kolom total_income nya

sorted(df.loc[df['total_income'].isna(), 'education'].unique())

["BACHELOR'S DEGREE",
 "Bachelor's Degree",
 'PRIMARY EDUCATION',
 'Primary Education',
 'SECONDARY EDUCATION',
 'SOME COLLEGE',
 'Secondary Education',
 'Some College',
 "bachelor's degree",
 'primary education',
 'secondary education',
 'some college']

In [23]:
# Menampilkan unique value di kolom education

sorted(df['education'].unique())

["BACHELOR'S DEGREE",
 "Bachelor's Degree",
 'GRADUATE DEGREE',
 'Graduate Degree',
 'PRIMARY EDUCATION',
 'Primary Education',
 'SECONDARY EDUCATION',
 'SOME COLLEGE',
 'Secondary Education',
 'Some College',
 "bachelor's degree",
 'graduate degree',
 'primary education',
 'secondary education',
 'some college']

In [24]:
# sample distribusi data dari salah satu value di kolom education yang terdapat missing value pada kolom total_income nya

df.loc[(df['education']=='primary education') & (df['total_income'].isna())]

children  days_employed  dob_years          education  education_id  \
1303          1            NaN         70  primary education             3   
1510          0            NaN         52  primary education             3   
1520          0            NaN         53  primary education             3   
4388          0            NaN         39  primary education             3   
4585          1            NaN         58  primary education             3   
7314          1            NaN         45  primary education             3   
8142          0            NaN         64  primary education             3   
10998         0            NaN         34  primary education             3   
11758         0            NaN         58  primary education             3   
13031         0            NaN         37  primary education             3   
13074         3            NaN         59  primary education             3   
13493         0            NaN         63  primary education             3   
15074         0            NaN         62  primary education             3   
17014         0            NaN         39  primary education             3   
17430         0            NaN         60  primary education             3   
17696         1            NaN         45  primary education             3   
18852         0            NaN         26  primary education             3   
20167         0            NaN         36  primary education             3   
20283         0            NaN         61  primary education             3   

           family_status  family_status_id gender    income_type  debt  \
1303   civil partnership                 1      F       employee     0   
1510   civil partnership                 1      M       employee     1   
1520             married                 0      F       employee     0   
4388             married                 0      F       employee     0   
4585   civil partnership                 1      F       employee     0   
7314             married                 0      M       employee     0   
8142   civil partnership                 1      F  civil servant     0   
10998            married                 0      F       employee     0   
11758            married                 0      M        retiree     0   
13031  civil partnership                 1      F       business     1   
13074            married                 0      M       employee     0   
13493            married                 0      F        retiree     0   
15074  civil partnership                 1      F       employee     0   
17014            married                 0      F  civil servant     0   
17430    widow / widower                 2      F        retiree     0   
17696            married                 0      F        retiree     0   
18852          unmarried                 4      M       employee     0   
20167          unmarried                 4      M       employee     0   
20283    widow / widower                 2      F        retiree     0   

       total_income                                   purpose  
1303            NaN  transactions with commercial real estate  
1510            NaN              construction of own property  
1520            NaN          transactions with my real estate  
4388            NaN                       building a property  
4585            NaN                      housing transactions  
7314            NaN                         profile education  
8142            NaN                         to have a wedding  
10998           NaN                           buy real estate  
11758           NaN          transactions with my real estate  
13031           NaN                          wedding ceremony  
13074           NaN                  getting higher education  
13493           NaN                              to own a car  
15074           NaN                  real estate transactions  
17014           NaN  transactions with commercial real estate  
17430

In [25]:
# Menampilkan unique value di kolom children yang terdapat missing value pada kolom total_income nya

sorted(df.loc[df['total_income'].isna(), 'children'].unique())

[-1, 0, 1, 2, 3, 4, 5, 20]

In [26]:
# Menampilkan unique value di kolom children

sorted(df['children'].unique())

[-1, 0, 1, 2, 3, 4, 5, 20]

In [27]:
# sample distribusi data dari salah satu value di kolom education yang terdapat missing value pada kolom total_income nya

df.loc[(df['children']==1) & (df['total_income'].isna())]

children  days_employed  dob_years            education  education_id  \
72            1            NaN         32    bachelor's degree             0   
94            1            NaN         34    bachelor's degree             0   
189           1            NaN         30  secondary education             1   
205           1            NaN         31    bachelor's degree             0   
220           1            NaN         23         some college             2   
...         ...            ...        ...                  ...           ...   
21407         1            NaN         36  secondary education             1   
21432         1            NaN         38         some college             2   
21463         1            NaN         35    bachelor's degree             0   
21495         1            NaN         50  secondary education             1   
21502         1            NaN         42  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
72               married                 0      M  civil servant     0   
94     civil partnership                 1      F       business     0   
189            unmarried                 4      F       employee     0   
205              married                 0      F       employee     0   
220    civil partnership                 1      F       business     0   
...                  ...               ...    ...            ...   ...   
21407            married                 0      F       business     0   
21432          unmarried                 4      F       employee     0   
21463  civil partnership                 1      M       employee     0   
21495  civil partnership                 1      F       employee     0   
21502            married                 0      F       employee     0   

       total_income                                   purpose  
72              NaN  transactions with commercial real estate  
94              NaN                          having a wedding  
189             NaN                              to own a car  
205             NaN                  buying a second-hand car  
220             NaN                         to have a wedding  
...             ...                                       ...  
21407           NaN                    building a real estate  
21432           NaN                      housing transactions  
21463           NaN                          having a wedding  
21495           NaN                          wedding ceremony  
21502           NaN                    building a real estate  

[475 rows x 12 columns]

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Didalam konteks membandingkan distribusi bisa dibandingkan proporsi pada nilai-nilai kategorik seperti `income_type` di dua dataset yang berbeda, satu dataset yang telah terfilter missing value saja dan keseluruhan dataset. Sehingga diketahui apakah terdapat perbedaan pola pada kedua dataset.
</div>

**Kesimpulan menengah**

Setelah membandingkan distribusi dataset asli dan dataset yang terfilter berdasarkan nilai kosong, didapati bahwa:
- terdapat 1 nilai 'XNA' pada kolom `gender` di dataset yang hilang (bisa jadi karena salah proses input),
- terdapat nilai 'graduate degree' pada kolom `education` di dataset yang tidak hilang , sedangkan di dataset yang hilang tidak
- terdapat nilai 'paternity / maternity leave' dan 'unemployed' pada kolom`income_type` di dataset yang tidak hilang , sedangkan di dataset yang hilang tidak
- tidak terdapat perbedaan unique value pada kolom `family_status`
- tidak terdapat perbedaan unique value pada kolom `children`

**Kemungkinan penyebab hilangnya nilai dalam data**

Kemungkinan hilangnya nilai dalam data dikarenakan kesalahan pada sistem saat proses mendapatkan data pada kolom `days_employed` dan `total_income`. 

**Kesimpulan**

Terdapat kesamaan jumlah baris yang hilang pada kolom `days_employed` dan `total_income`. Setelah membandingkan distribusi antara data yang hilang dan utuh, hilangnya nilai dalam data diperkirakan karena kesalahan pada sistem saat proses mendapatkan data untuk kolom `days_employed` dan `total_income`.

Nilai yang hilang dapat diatasi dengan mengisnya dengan mean, median, ataupun modus kolom, tergantung pada distribusi nilai di kolom tersebut.

Langkah selanjutnya adalah mengatasi nilai yang bermasalah pada data (duplikat, register yang berbeda, data lama yang salah, dan nilai yang hilang)

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Instruksi yang ada hanya akan membantu memberi guideline bagaiman Chulsum seharusnya memberi penjelasan di tiap langkahnya pada project ini maka dari itu pastikan penjelasan dapat dipahami oleh orang lain.
</div>

## Transformasi data

In [28]:
# Menampilkan unique value pada kolom education untuk memeriksa ejaan apa yang perlu diperbaiki

sorted(df['education'].unique())

["BACHELOR'S DEGREE",
 "Bachelor's Degree",
 'GRADUATE DEGREE',
 'Graduate Degree',
 'PRIMARY EDUCATION',
 'Primary Education',
 'SECONDARY EDUCATION',
 'SOME COLLEGE',
 'Secondary Education',
 'Some College',
 "bachelor's degree",
 'graduate degree',
 'primary education',
 'secondary education',
 'some college']

In [29]:
# menghitung jumlah duplikat

df.duplicated().sum()

54

In [30]:
# Memperbaiki ejaan di kolom education 

df['education'] = df['education'].str.lower()

In [31]:
# menghitung jumlah duplikat

df.duplicated().sum()

71

In [32]:
# Memeriksa semua nilai di kolom untuk memastikan kita telah memperbaikinya

sorted(df['education'].unique())

["bachelor's degree",
 'graduate degree',
 'primary education',
 'secondary education',
 'some college']

Pada kolom education terdapat nilai dengan ejaan yang sama tetapi dengan penulisan kapital yang berbeda, maka dari itu dilakukan penyamaan nilai dengan cara mengecilkan huruf. Pembetulan ejaan tersebut berefek pada proses pendeteksian duplikat baris.

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Bisa ditambahkan pengecekan duplikat sebelum dan sesudah penggunaan `lower()` pada kolom `education` dan berikan penjelasan terkait dampaknya.

</div>

[Periksa data kolom `children`]

In [33]:
# Melihat distribusi nilai pada kolom `children`

df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [34]:
# menghitung persentase nilai bermasalah

round(df.loc[(df['children']==-1), 'children'].count() / len(df) * 100, 2)

0.22

Hal yang tidak normal pada kolom `children`:
- Terdapat nilai negatif (tidak rasional mengingat deskripsi dari data adalah anak)
- Terdapat nilai 20 dengan jumlah 20 user. Nilai tersebut memiliki gap yang cukup jauh dari nilai yang lain. Ada kemungkinan kesalahan saat proses input dari pengguna (typo), yang seharusnya tertulis 2 jadi 20

Ketidak normalah pada nilai tersebut dapat diperbaiki dengan cara:
- mengubah nilai negatif menjadi positif
- mengganti nilai 20 menjadi 2

In [35]:
# Mengubah nilai negatif menjadi positif

df['children'] = df['children'].abs()

In [36]:
# memeriksa kolom nilai `children`

df['children'].value_counts()

0     14149
1      4865
2      2055
3       330
20       76
4        41
5         9
Name: children, dtype: int64

In [37]:
# Mengganti nilai 20 menjadi 2

df.loc[df['children']==20, 'children'] = 2

In [38]:
# Periksa kembali kolom `children` untuk memastikan semua telah diperbaiki

df['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

20 Anak sebanyak 76 user merupakan sesuatu yang tidak biasa, identifikasi hal tersebut dan dapat diasumsikan merupakan salah input yang seharusnya 2 anak atau tidak.
</div>

In [39]:
# Memeriksa kolom days_employed

df['days_employed'].describe()

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

Hal yang tidak normal pada kolom `days_employed`:
- terdapat nilai negatif (tidak rasional mengingat deskripsi dari data adalah lama nasabah bekerja dalam hari)
- terdapat nilai yang terlalu jauh (bila dianggap usia aktif bekerja selama 60 tahun, berarti max 60*360 = 21600 hari)

In [40]:
# Menampilkan data yang bermasalah (value pada kolom days_employed bernilai negatif)

df.loc[df['days_employed'] < 0]

children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  secondary education             1   
3             3   -4124.747207         32  secondary education             1   
5             0    -926.185831         27    bachelor's degree             0   
...         ...            ...        ...                  ...           ...   
21519         1   -2351.431934         37      graduate degree             4   
21520         1   -4529.316663         43  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
5      civil partnership                 1      M    business     0   
...                  ...               ...    ...         ...   ...   
21519           divorced                 3      M    employee     0   
21520  civil partnership                 1      F    business     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                     purpose  
0         40620.102       purchase of the house  
1         17932.802                car purchase  
2         23341.752       purchase of the house  
3         42820.568     supplementary education  
5         40922.170       purchase of the house  
...             ...                         ...  
21519     18551.846  buy commercial real estate  
21520     35966.698        housing transactions  
21522     14347.610                    property  
21523     39054.888           buying my own car  
21524     13127.587                to buy a car  

[15906 rows x 12 columns]

In [41]:
# Menampilkan data yang bermasalah (value pada kolom days_employed terlalu jauh)

df.loc[df['days_employed'] > 21600]

children  days_employed  dob_years            education  education_id  \
4             0  340266.072047         53  secondary education             1   
18            0  400281.136913         53  secondary education             1   
24            1  338551.952911         57  secondary education             1   
25            0  363548.489348         67  secondary education             1   
30            1  335581.668515         62  secondary education             1   
...         ...            ...        ...                  ...           ...   
21505         0  338904.866406         53  secondary education             1   
21508         0  386497.714078         62  secondary education             1   
21509         0  362161.054124         59    bachelor's degree             0   
21518         0  373995.710838         59  secondary education             1   
21521         0  343937.404131         67  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
4      civil partnership                 1      F     retiree     0   
18       widow / widower                 2      F     retiree     0   
24             unmarried                 4      F     retiree     0   
25               married                 0      M     retiree     0   
30               married                 0      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21505  civil partnership                 1      M     retiree     0   
21508            married                 0      M     retiree     0   
21509            married                 0      M     retiree     0   
21518            married                 0      F     retiree     0   
21521            married                 0      F     retiree     0   

       total_income                                   purpose  
4         25378.572                         to have a wedding  
18         9091.804                  buying a second-hand car  
24        46487.558  transactions with commercial real estate  
25         8818.041                           buy real estate  
30        27432.971  transactions with commercial real estate  
...             ...                                       ...  
21505     12070.399                         to have a wedding  
21508     11622.175                                  property  
21509     11684.650                  real estate transactions  
21518     24618.344                         purchase of a car  
21521     24959.969                         purchase of a car  

[3445 rows x 12 columns]

In [42]:
# menghitung persentase nilai yang bermasalah pada kolom days_employed

round(df.loc[(df['days_employed'] < 0) | (df['days_employed'] > 21600), 'days_employed'].count() / len(df['days_employed']) * 100, 2)

89.9

Persentasi nilai yang bermasalah sangat tinggi (89.9%). Hal ini bisa dikarenakan beberapa masalah teknis. Untuk memperbaiki nilai tersebut, hal yang dapat dilakukan yaitu:
- mengubah nilai negatif menjadi positif (dijadikan absolute)
- mengubah nilai yang terlalu jauh (dengan patokan usia aktif bekerja selama 60 tahun, berarti max 60*360 = 21600 hari) menjadi median atau mean (karena tipe data berupa numerik)

In [43]:
# mengubah nilai negatif menjadi positif

df['days_employed'] = df['days_employed'].abs()

In [44]:
# memeriksa data yang telah diubah

df.loc[df['days_employed'] < 0]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

In [45]:
# memeriksa nilai mean dan median untuk menentukan pengganti nilai yang terlalu jauh

df['days_employed'].describe()

count     19351.000000
mean      66914.728907
std      139030.880527
min          24.141633
25%         927.009265
50%        2194.220567
75%        5537.882441
max      401755.400475
Name: days_employed, dtype: float64

In [46]:
# mengubah nilai yang terlalu jauh menjadi median

df.loc[(df['days_employed'] > 21600), 'days_employed'] = df['days_employed'].median()

Alasan dari pemilihan median sebagai pengganti nilai yang terlampau jauh adalah karena penyebaran data tidak merata sehingga letak mean jauh dari median pada deskripsi data di atas. Hal tersebut menyebabkan mean tidak dapat mewakili keseluruhan data 

In [47]:
# memastikan nilai telah diperbaiki

df['days_employed'].describe()

count    19351.000000
mean      2324.746073
std       2089.961651
min         24.141633
25%        927.009265
50%       2194.220567
75%       2747.423625
max      18388.949901
Name: days_employed, dtype: float64

In [48]:
# Memeriksa kolom dob_years untuk nilai yang mencurigakan dan hitung persentasenya

sorted(df['dob_years'].unique())

[0,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75]

Terdapat nilai 0 pada kolom `dob_years` (nilai tersebut tidak rasional mengingat deskripsi dari kolom tersebut adalah usia nasabah)

Karena data bersifat numerik, nilai yang bermasalah dapat diselesaikan dengan menggantinya menjadi mean atau median (tergantung persebaran data pada kolom)

In [49]:
# memeriksa nilai mean dan median untuk menentukan pengganti nilai 0

df['dob_years'].describe()

count    21525.000000
mean        43.293380
std         12.574584
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

In [50]:
df.loc[df['dob_years']==0, 'dob_years'] = df['dob_years'].mean()

Alasan digunakannya mean sebagai pengganti nilai 0 adalah karena penyebaran merata sehingga letak mean tidak terlalu jauh dengan median. Hal tersebut menyebabkan mean dapat mewakili keseluruhan data 

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Akan lebih baik diberi penjelasan mengenai alasan dalam penentuan nilai imputasi menggunakan mean median atau nilai lain.
</div>

In [51]:
# memastikan nilai telah diperbaiki

sorted(df['dob_years'].unique())

[19.0,
 20.0,
 21.0,
 22.0,
 23.0,
 24.0,
 25.0,
 26.0,
 27.0,
 28.0,
 29.0,
 30.0,
 31.0,
 32.0,
 33.0,
 34.0,
 35.0,
 36.0,
 37.0,
 38.0,
 39.0,
 40.0,
 41.0,
 42.0,
 43.0,
 43.29337979094077,
 44.0,
 45.0,
 46.0,
 47.0,
 48.0,
 49.0,
 50.0,
 51.0,
 52.0,
 53.0,
 54.0,
 55.0,
 56.0,
 57.0,
 58.0,
 59.0,
 60.0,
 61.0,
 62.0,
 63.0,
 64.0,
 65.0,
 66.0,
 67.0,
 68.0,
 69.0,
 70.0,
 71.0,
 72.0,
 73.0,
 74.0,
 75.0]

In [52]:
# Melihat unique value pada kolom family_status

df['family_status'].unique()

array(['married', 'civil partnership', 'widow / widower', 'divorced',
       'unmarried'], dtype=object)

In [53]:
# Melihat deskripsi dari kolom `family_status`

df['family_status'].describe()

count       21525
unique          5
top       married
freq        12380
Name: family_status, dtype: object

Setelah pengecekan pada kolom `family_status`, tidak ditemukan adanya keanehan pada data

In [54]:
# Melihat nilai di kolom gender

df['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

Terdapat 1 nilai 'XNA' pada kolom `gender`. Hal tersebut bisa terjadi karena kesalahan pada sistem saat proses input data

Masalah tersebut dapat diatasi dengan mengganti nilai 'XNA' menjadi modus data (karena data bersifat kategorik)

In [55]:
# menampilkan deskripsi dari kolom ender

df['gender'].describe()

count     21525
unique        3
top           F
freq      14236
Name: gender, dtype: object

In [56]:
# mengubah nilai 'XNA' menjadi modus

df.loc[df['gender']=='XNA', 'gender'] = df['gender'].mode()[0]

In [57]:
# memastikan nilai telah diperbaiki

df['gender'].value_counts()

F    14237
M     7288
Name: gender, dtype: int64

In [58]:
# Mari kita lihat nilai dalam kolom

df['income_type'].describe()

count        21525
unique           8
top       employee
freq         11119
Name: income_type, dtype: object

Setelah melakukan pengecekan pada kolom `income_type`, diketahui bahwa tidak terdapat nilai yang bermasalah

Karena value yang bermasalah sudah diselesaikan, bisa dilanjutkan dengan proses penghapusan data duplikat tanpa takut merusak nilai dari data

In [59]:
# Memeriksa duplikat

df.duplicated().sum()

71

In [60]:
# melakukan penghapusan duplikat

df = df.drop_duplicates().reset_index(drop=True) 

In [61]:
# Memeriksa apakah data masih memiliki duplikat

df.duplicated().sum()

0

In [62]:
# Periksa ukuran dataset setelah proses manipulasi

df.shape

(21454, 12)

Kini dataset telah bersih dari nilai duplikat, register yang berbeda, maupun data lama yang salah

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Perhatikan instruksi yang ada dalam memberikan penjelasan di tahapan pengerjaan proyek ini.
</div>

# Bekerja dengan nilai yang hilang

In [63]:
# Temukan dictionary 

sorted(df['dob_years'].unique())

[19.0,
 20.0,
 21.0,
 22.0,
 23.0,
 24.0,
 25.0,
 26.0,
 27.0,
 28.0,
 29.0,
 30.0,
 31.0,
 32.0,
 33.0,
 34.0,
 35.0,
 36.0,
 37.0,
 38.0,
 39.0,
 40.0,
 41.0,
 42.0,
 43.0,
 43.29337979094077,
 44.0,
 45.0,
 46.0,
 47.0,
 48.0,
 49.0,
 50.0,
 51.0,
 52.0,
 53.0,
 54.0,
 55.0,
 56.0,
 57.0,
 58.0,
 59.0,
 60.0,
 61.0,
 62.0,
 63.0,
 64.0,
 65.0,
 66.0,
 67.0,
 68.0,
 69.0,
 70.0,
 71.0,
 72.0,
 73.0,
 74.0,
 75.0]

### Memperbaiki nilai yang hilang di `total_income`

Kolom yang yang memiliki nilai hilang adalah `total_income` dan `days_employed`. 

Pertama, kategorikan usia untuk memudahkan dalam hal menghitung total_income nanti. Pengkategorian dilakukan berdasarkan [situs](https://www.statcan.gc.ca/en/concepts/definitions/age2) ini

In [64]:
# Mari menulis fungsi untuk menghitung kategori usia

def age_category(age):
    if age <= 14:
        return 'Children'
    elif (age > 14) & (age <= 24):
        return 'Youth'
    elif (age > 24) & (age <= 64):
        return 'Adults'
    elif age > 64:
        return 'Senior'

<div class="alert alert-warning">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Untuk menghindari subjektifitas dan tidak adanya referensi yang digunakan dalam pengelompokan, kategori dapat dinamai sesuai dengan batas atas dan bawah pada tiap kelompok misa, `sampai 24 tahun` dan `64 tahun keatas`.

</div>

In [65]:
# melakukan pengujian terhadap fungsi

age_category(12.0)

'Children'

In [66]:
# Membuat kolom baru berdasarkan fungsi

df['age_category'] = df['dob_years'].apply(age_category)

In [67]:
# Memeriksa bagaimana nilai di dalam kolom baru

df['age_category'].value_counts()

Adults    19684
Senior      895
Youth       875
Name: age_category, dtype: int64

In [68]:
# Membuat tabel tanpa nilai yang hilang

not_null = df.loc[~(df['days_employed'].isna()) & ~(df['total_income'].isna())]
not_null

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028       42.0    bachelor's degree             0   
1             1    4024.803754       36.0  secondary education             1   
2             0    5623.422610       33.0  secondary education             1   
3             3    4124.747207       32.0  secondary education             1   
4             0    2194.220567       53.0  secondary education             1   
...         ...            ...        ...                  ...           ...   
21449         1    4529.316663       43.0  secondary education             1   
21450         0    2194.220567       67.0  secondary education             1   
21451         1    2113.346888       38.0  secondary education             1   
21452         3    3112.481705       38.0  secondary education             1   
21453         2    1984.507589       40.0  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21449  civil partnership                 1      F    business     0   
21450            married                 0      F     retiree     0   
21451  civil partnership                 1      M    employee     1   
21452            married                 0      M    employee     1   
21453            married                 0      F    employee     0   

       total_income                  purpose age_category  
0         40620.102    purchase of the house       Adults  
1         17932.802             car purchase       Adults  
2         23341.752    purchase of the house       Adults  
3         42820.568  supplementary education       Adults  
4         25378.572        to have a wedding       Adults  
...             ...                      ...          ...  
21449     35966.698     housing transactions       Adults  
21450     24959.969        purchase of a car       Senior  
21451     14347.610                 property       Adults  
21452     39054.888        buying my own car       Adults  
21453     13127.587             to buy a car       Adults  

[19351 rows x 13 columns]

In [69]:
# menampilkan rata-rata dari kolom total_income

not_null['total_income'].mean()

26787.568354658673

In [70]:
# menampilkan median dari kolom total_income

not_null['total_income'].median()

23202.87

In [71]:
# menampilkan deskripsi dari kolom total_income

df['total_income'].describe()

count     19351.000000
mean      26787.568355
std       16475.450632
min        3306.762000
25%       16488.504500
50%       23202.870000
75%       32549.611000
max      362496.645000
Name: total_income, dtype: float64

Setelah melakukan pengecekan pada kolom `total_income`, saya memutuskan untuk menggunakan mean sebagai pengganti nilai yang kosong pada kolom tersebut. Alasan dari digunakannya mean adalah karena tipe data dari kolom tersebut adalah numerik. Selain itu, jarak antara mean dan mediannya tidak terlalu jauh dan persebaran datanya merata jadi nilai mean dapat merepresentasikan keseluruhan data

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Sebaiknya di elaborate lagi terkait standard deviasi 20% ini.
</div>

In [72]:
#  Membuat fungsi yang akan gunakan untuk mengisi nilai yang hilang

def data_imputation(data, column_grouping, column_selected):
    group = data[column_grouping].unique()

    for rule in group:
        mean = data.loc[(data[column_grouping]==rule) & ~(data[column_selected].isna()), column_selected].mean()
        data.loc[(data[column_grouping]==rule) & (data[column_selected].isna()), column_selected] = mean
        
    return data

In [73]:
# mengimplementasikan fungsi pada proses pengisian kolom total_income dengan referensi kolom age_category

df = data_imputation(df, column_grouping='age_category', column_selected='total_income')

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Ketika imputasi `total_income` menggunakan satu nilai maka semua missing value akan berubah menjadi nilai yang sama. Baik pada umur, type income, sampai pendidikan tidak akan memiliki perbedaan dari `total_income`. Sebaiknya digunakan paling tidak satu kolom sebagai referensi dalam melakukan imputasi ini.
</div>

<div class="alert alert-success">
<b>Reviewer's comment v2</b> <a class="tocSkip"></a>

Walaupun fungsi yang dibuat belum sepenuhnya efektif namun ini adalah perbaikan yang cukup bagus mengingat sudah sesuai dengan yang dibutuhkan.
</div>

In [74]:
# Memeriksa apakah fungsi berjalan

df.isna().sum()

children               0
days_employed       2103
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
age_category           0
dtype: int64

In [75]:
# Memeriksa jumlah entri di kolom

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21454 non-null  float64
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   debt              21454 non-null  int64  
 10  total_income      21454 non-null  float64
 11  purpose           21454 non-null  object 
 12  age_category      21454 non-null  object 
dtypes: float64(3), int64(4), object(6)
memory usage: 2.1+ MB


###  Memperbaiki nilai di `days_employed`

In [76]:
# menampilkan distribusi median dari `days_employed` berdasarkan parameter age_category

pd.pivot_table(
    df,
    values='days_employed',
    index=['age_category', 'income_type'],
    aggfunc=[np.median]
)

median
                                         days_employed
age_category income_type                              
Adults       business                      1618.911339
             civil servant                 2776.347655
             employee                      1651.060608
             entrepreneur                   520.848083
             paternity / maternity leave   3296.759962
             retiree                       2194.220567
             unemployed                    2194.220567
Senior       business                      1928.123732
             civil servant                 3339.663548
             employee                      3240.640456
             retiree                       2194.220567
Youth        business                       695.968951
             civil servant                  922.372533
             employee                       769.758975
             retiree                       2194.220567
             student                        578.751554

In [77]:
# menampilkan distribusi rata-rata dari `days_employed` berdasarkan age_category

pd.pivot_table(
    df,
    values='days_employed',
    index=['age_category', 'income_type'],
    aggfunc=[np.mean]
)

mean
                                         days_employed
age_category income_type                              
Adults       business                      2175.365499
             civil servant                 3469.888219
             employee                      2388.724641
             entrepreneur                   520.848083
             paternity / maternity leave   3296.759962
             retiree                       2194.220567
             unemployed                    2194.220567
Senior       business                      3292.768472
             civil servant                 4377.924812
             employee                      4366.938943
             retiree                       2194.220567
Youth        business                       793.557137
             civil servant                  975.347977
             employee                       873.864193
             retiree                       2194.220567
             student                        578.751554

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Karena sudah dihitung nilai rata-rata pada tiap age_group dan type_income seharusnya nilai-nilai ini bisa digunakan sebagai imputasi pada masing-masing group hasil kombinasi antara age_group dan type_income baik pada total_income maupun days_employed.
</div>

In [78]:
# menampilkan deskripsi dari kolom total_income

df['days_employed'].describe()

count    19351.000000
mean      2324.746073
std       2089.961651
min         24.141633
25%        927.009265
50%       2194.220567
75%       2747.423625
max      18388.949901
Name: days_employed, dtype: float64

Setelah melakukan pengecekan pada kolom `days_employed`, saya memutuskan untuk menggunakan mean sebagai pengganti nilai yang kosong pada kolom tersebut. Alasan dari digunakannya mean adalah karena tipe data dari kolom tersebut adalah numerik. Selain itu, jarak antara mean dan mediannya tidak terlalu jauh dan persebaran datanya merata jadi nilai mean dapat merepresentasikan keseluruhan data.

Proses imputasi pada kolom `days_employed` dapat menggunakan fungsi yang sama seperti yang digunakan pada proses imputasi di kolom `total_income` (data_imputation). Hal ini karena proses imputasi kebetulan sama-sama menggunakan mean dan parameter yang digunakan pada fungsi tersebut juga sesuai dengan proses imputasi pada kolom `days_employed`

In [79]:
# mengimplementasikan fungsi data_imputation pada proses pengisian kolom days_employed dengan referensi kolom age_category

df = data_imputation(df, column_grouping='age_category', column_selected='days_employed')

In [80]:
# memeriksa fungsi bekerja

df.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
age_category        0
dtype: int64

In [81]:
# Memeriksa jumlah entri di kolom

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  int64  
 1   days_employed     21454 non-null  float64
 2   dob_years         21454 non-null  float64
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   debt              21454 non-null  int64  
 10  total_income      21454 non-null  float64
 11  purpose           21454 non-null  object 
 12  age_category      21454 non-null  object 
dtypes: float64(3), int64(4), object(6)
memory usage: 2.1+ MB


## Pengkategorian Data

In [82]:
# Menampilkan kolom yang dipilih untuk pengkategorian (debt)

df['debt']

0        0
1        0
2        0
3        0
4        0
        ..
21449    0
21450    0
21451    1
21452    1
21453    0
Name: debt, Length: 21454, dtype: int64

In [83]:
# memeriksa nilai unik pada kolom debt 

df['debt'].unique()

array([0, 1])

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Hindari penggunaan for loop untuk langkah ini.
</div>

In [84]:
# fungsi untuk mengkategorikan data berdasarkan nilai debtnya

def replace_debt(val):
    if val==1:
        return 'gagal bayar'
    elif val==0:
        return 'tidak gagal bayar'

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Coba pahami lagi terkait deskripsi data pada penjelasan awal terkait pernah tidaknya seseorang gagal bayar dalam berhutang.
</div>

In [85]:
# mengimplementasikan fungsi replace_debt 

df['debt_meaning'] = df['debt'].apply(replace_debt)
df.head()

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028       42.0    bachelor's degree             0   
1         1    4024.803754       36.0  secondary education             1   
2         0    5623.422610       33.0  secondary education             1   
3         3    4124.747207       32.0  secondary education             1   
4         0    2194.220567       53.0  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose age_category       debt_meaning  
0    purchase of the house       Adults  tidak gagal bayar  
1             car purchase       Adults  tidak gagal bayar  
2    purchase of the house       Adults  tidak gagal bayar  
3  supplementary education       Adults  tidak gagal bayar  
4        to have a wedding       Adults  tidak gagal bayar

In [86]:
# Menghitung setiap nilai debt_meaning untuk melihat pendistribusian

df['debt_meaning'].value_counts()

tidak gagal bayar    19713
gagal bayar           1741
Name: debt_meaning, dtype: int64

In [87]:
# Menampilkan komom yang dipilih untuk pengkategorian (total_income)

df['total_income']

0        40620.102
1        17932.802
2        23341.752
3        42820.568
4        25378.572
           ...    
21449    35966.698
21450    24959.969
21451    14347.610
21452    39054.888
21453    13127.587
Name: total_income, Length: 21454, dtype: float64

In [88]:
# memeriksa nilai unik pada kolom total_income

df['total_income'].unique()

array([40620.102, 17932.802, 23341.752, ..., 14347.61 , 39054.888,
       13127.587])

In [89]:
# Mendapatkan kesimpulan statistik untuk kolom total_income

df['total_income'].describe()

count     21454.000000
mean      26788.328876
std       15653.145234
min        3306.762000
25%       17219.817250
50%       24671.402000
75%       31330.237250
max      362496.645000
Name: total_income, dtype: float64

Pengkategorian pendapatan dilakukan berdasarkan [situs](https://blogs.worldbank.org/opendata/new-world-bank-country-classifications-income-level-2021-2022) ini

In [90]:
# Membuat fungsi untuk mengkategorikan pemasukan

def income_category(val):
    if val < 1035:
        return 'low income' 
    elif (val >= 1035) & (val <= 4045):
        return 'lower-middle income'
    elif (val >= 4046) & (val <= 12535):
        return 'upper-middle income'
    elif val > 12535:
        return 'high income'

In [91]:
# Mengaplikasikan fungsi ke kolom baru

df['income_category'] = df['total_income'].apply(income_category)

In [92]:
# menghitung nilai di kolom baru

df['income_category'].value_counts()

high income            19345
upper-middle income     2100
lower-middle income        9
Name: income_category, dtype: int64

In [93]:
# Menampilkan komom yang dipilih untuk pengkategorian (purpose)

df['purpose']

0          purchase of the house
1                   car purchase
2          purchase of the house
3        supplementary education
4              to have a wedding
                  ...           
21449       housing transactions
21450          purchase of a car
21451                   property
21452          buying my own car
21453               to buy a car
Name: purpose, Length: 21454, dtype: object

In [94]:
# memeriksa nilai unik pada kolom purpose

df['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

In [95]:
# mengkategorikan purpose

df.loc[df['purpose'].isin(['purchase of the house', 'housing transactions', 'purchase of the house for my family', 'housing', 'purchase of my own house', 'housing renovation']), 'purpose'] = 'house'
df.loc[df['purpose'].isin(['car purchase', 'second-hand car purchase', 'buying a second-hand car', 'buying my own car', 'cars', 'car', 'to own a car', 'purchase of a car', 'to buy a car']), 'purpose'] = 'car'
df.loc[df['purpose'].isin(['supplementary education', 'education', 'to become educated', 'getting an education', 'to get a supplementary education', 'getting higher education', 'profile education', 'university education', 'going to university']), 'purpose'] = 'education'
df.loc[df['purpose'].isin(['buy real estate', 'buy commercial real estate', 'buy residential real estate', 'transactions with commercial real estate', 'building a real estate', 'transactions with my real estate', 'real estate transactions', 'buying property for renting out']), 'purpose'] = 'realestate'
df.loc[df['purpose'].isin(['construction of own property', 'property', 'building a property',]), 'purpose'] = 'property'
df.loc[df['purpose'].isin(['to have a wedding', 'having a wedding', 'wedding ceremony']), 'purpose'] = 'wedding'

In [96]:
df['purpose'].value_counts()

realestate    5115
car           4306
education     4013
house         3809
wedding       2324
property      1887
Name: purpose, dtype: int64

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Hindari penggunaan for loop untuk langkah ini.
</div>

## Memeriksa Hipotesis


**Apakah terdapat korelasi antara memiliki anak dengan membayar kembali tepat waktu?**

In [97]:
# Memeriksa data children dan debt

df[['children', 'debt']]

children  debt
0             1     0
1             1     0
2             0     0
3             3     0
4             0     0
...         ...   ...
21449         1     0
21450         0     0
21451         1     1
21452         3     1
21453         2     0

[21454 rows x 2 columns]

In [98]:
# Menghitung default rate berdasarkan jumlah anak

pivot_children = pd.pivot_table(df, index='children', columns='debt',  values='education', aggfunc='count')
pivot_children.reset_index(inplace=True)
pivot_children['percentage'] = round(pivot_children[1] / (pivot_children[0] + pivot_children[1]) * 100, 2)
pivot_children

debt  children        0       1  percentage
0            0  13028.0  1063.0        7.54
1            1   4410.0   445.0        9.17
2            2   1926.0   202.0        9.49
3            3    303.0    27.0        8.18
4            4     37.0     4.0        9.76
5            5      9.0     NaN         NaN

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Hindari penggunaan multipel output pada satu sel. Bisa digunakan pivot tabel untuk menggabung kedua output tersebut yang berisi gabungan antara proporsi dua kategori pada kolom debt terhadap kategori jumlah anak serta persentase yang pernah gagal bayar terhadap keseluruhan.
    
Penggunaan `mean` disini hanya kebetulan semata terkait mapping nilai 0 dan 1 terhadap kategori gagal bayar tidaknya.
</div>

<div class="alert alert-success">
<b>Reviewer's comment v2</b> <a class="tocSkip"></a>

Well-done!
</div>

**Kesimpulan**

Nasabah dengan 5 anak persentase gagal bayarnya paling rendah, tetapi jumlah datanya sangat sedikit. Nasabah yang tidak memiliki anak menempati persentase gagal bayarnya paling rendah ke 2, dilanjutkan dengan nasabah ber anak 1, 3, 2, kemudian 4. Ditemukan bahwa nasabah dengan jumlah anak yang lebih sedikit cenderung memiliki persentase gagal bayar yang lebih rendah.

**Apakah terdapat korelasi antara status keluarga dengan membayar kembali tepat waktu?**

In [99]:
# Memeriksa data family_status dan debt

print(df[['family_status', 'debt']])

           family_status  debt
0                married     0
1                married     0
2                married     0
3                married     0
4      civil partnership     0
...                  ...   ...
21449  civil partnership     0
21450            married     0
21451  civil partnership     1
21452            married     1
21453            married     0

[21454 rows x 2 columns]


In [100]:
# Menghitung default rate berdasarkan family_status

pivot_family = pd.pivot_table(df, index='family_status', columns='debt',  values='education', aggfunc='count')
pivot_family.reset_index(inplace=True)
pivot_family['percentage'] = round(pivot_family[1] / (pivot_family[0] + pivot_family[1]) * 100, 2)
pivot_family

debt      family_status      0    1  percentage
0     civil partnership   3763  388        9.35
1              divorced   1110   85        7.11
2               married  11408  931        7.55
3             unmarried   2536  274        9.75
4       widow / widower    896   63        6.57

**Kesimpulan**

Nasabah yang berstatus widow / widower memiliki persentase gagal bayar yang paling rendah, sedangkan nasabah dengan status unmarried memiliki persentase gagal bayar tertinggi. Berdasarkan tabel di atas, dapat disimpulkan bahwa nasabah yang sudah pernah menikah (married, divorced, dan widow / widower) memiliki memiliki persentase gagal bayar yang lebih kecil dibandingkan yang belum pernah menikah (civil partnership dan unmarried)

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Coba cek review pada pertanyaan pertama.
</div>

**Apakah terdapat korelasi antara tingkat pendapatan dengan membayar kembali tepat waktu?**

In [101]:
# memeriksa data income_category dan debt
 
df[['income_category', 'debt']]

income_category  debt
0         high income     0
1         high income     0
2         high income     0
3         high income     0
4         high income     0
...               ...   ...
21449     high income     0
21450     high income     0
21451     high income     1
21452     high income     1
21453     high income     0

[21454 rows x 2 columns]

In [102]:
# Menghitung default rate berdasarkan income_category

pivot_income = pd.pivot_table(df, index='income_category', columns='debt',  values='education', aggfunc='count')
pivot_income.reset_index(inplace=True)
pivot_income['percentage'] = round(pivot_income[1] / (pivot_income[0] + pivot_income[1]) * 100, 2)
pivot_income

debt      income_category      0     1  percentage
0             high income  17758  1587        8.20
1     lower-middle income      8     1       11.11
2     upper-middle income   1947   153        7.29

**Kesimpulan**

Nasabah dengan lower-middle income memiliki persentase gagal yang paling besar, tetapi datanya sangat sedikit. Dari tabel tersebut dapat diketahui bahwa semakin tinggi pendapatan nasabah, maka semakin rendah persentase gagal bayarnya

**Bagaimana tujuan kredit memengaruhi default rate?**

In [103]:
# memeriksa data income_category dan debt
 
df[['purpose', 'debt']]

purpose  debt
0          house     0
1            car     0
2          house     0
3      education     0
4        wedding     0
...          ...   ...
21449      house     0
21450        car     0
21451   property     1
21452        car     1
21453        car     0

[21454 rows x 2 columns]

In [104]:
# Menghitung default rate berdasarkan purpose

pivot_purpose = pd.pivot_table(df, index='purpose', columns='debt',  values='education', aggfunc='count')
pivot_purpose.reset_index(inplace=True)
pivot_purpose['percentage'] = round(pivot_purpose[1] / (pivot_purpose[0] + pivot_purpose[1]) * 100, 2)
pivot_purpose

debt     purpose     0    1  percentage
0            car  3903  403        9.36
1      education  3643  370        9.22
2          house  3553  256        6.72
3       property  1749  138        7.31
4     realestate  4727  388        7.59
5        wedding  2138  186        8.00

<div class="alert alert-danger">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Coba cek review pada pertanyaan pertama. Sebaiknya terkait purpose dapat dilakukan pengkategorian pada tahap sebelumnya.
</div>

**Kesimpulan**

Nasabah dengan tujuan 'rumah' memiliki persentase gagal bayar yang paling rendah, sedangkan nasabah dengan tujuan car  dan education memiliki kecenderungan gagal bayar yang tinggi. Dari data tersebut didapati bahwa nasabah dengan tujuan membelanjakan uangnya ke bidang bangunan (rumah, properti, atau realestate) memiliki persentase gagal bayar yang lebih kecil dibandingkan tujuan lain

# Kesimpulan Umum 

1. Eksplorasi data
Langkah pertama yaitu mengimport library, read data, serta memahami data. Hal ini bisa dilakukan dengan cara membaca dictionary data serta menggunakan beberapa metode bawaan python seperti info(), describe(), dll. Setelah mengecek data, ternyata terdapat nilai kosong pada kolom `days_employed` dan `total_income`. Melihat dari distribusi nilai yang hilang, dugaan sementara yaitu nilai kosong tersebut berpola. Langkah selanjutnya yaitu memeriksa persentase nilai yang hilang dengan seluruh kumpulan data, kemudian membandingkan value dan distribusi antara data seluruh dataset dengan data yang terfilter berdasarkan nilai kosong. Berdasarkan proses tersebut, didapati bahwa terdapat perbedaan di beberapa kolom dan dapat disimpulkan bahwa hilangnya nilai dalam data dikarenakan kesalahan pada sistem saat proses mendapatkan data pada kolom `days_employed` dan `total_income`. 
2. Transformasi data
Langkah selanjutnya adalah mencari dan mengatasi nilai yang bermasalah pada data (register yang berbeda, nilai yang tidak rasional, ataupun data lama yang salah). Nilai yang bermasalah dapat substitusi dengan mean, median, modus, atau beberapa nilai khusus tergantung pada masalah dan tipe data yang sedang dihadapi. Jika data bersifat numerik, nilai yang bermasalah dapat diselesaikan dengan menggantinya menjadi mean atau median (tergantung persebaran data pada kolom), sedangkan jika data bersifat kategorikal dapat diganti dengan modus. Setelah nilai yang bermasalah terselesaikan, bisa dilanjutkan dengan proses penghapusan data duplikat tanpa takut merusak nilai dari data.
3. Mengolah missing value
Proses selanjutnya adalah mengisi data yang hilang (missing value). Missing value dapat substitusi dengan mean, median, atau modus, penyebaran dan tipe data yang sedang dihadapi. Kolom yang yang memiliki missing value adalah `total_income` dan `days_employed`. Saya melakukan pengkategorian data usia terlebih dahulu untuk memudahkan dalam hal menghitung total_income nantinya. Saya membuat function data_imputation yang berfungsi untuk memudahkan dalam proses pengisian missing value berdasarkan data kategorikal di kolom lain.
4. Pengkategorian data
Proses ini dilakukan untuk menjawab pertanyaan dan menguji hipotesis. Saya mengkategorikan `debt` kedalam `debt_meaning`, `total_income` kedalam `income_category`, dan `purpose`. Pertama tampilkan kolom yang dipilih untuk pengkategorian dan melakukan pengecekan kepada unique values-nya. Kemudian, buat fungsi untuk pengkategoriannya dan implementasikan fungsinya.

    Tidak lupa setelah melakukan perubahan pada data di tiap kegiatan (eksplorasi data, transformasi data, mengolah missing value, pengkategorian data) selalu dilakukan pengecekan untuk melihat hasil dan mencegah kesalahan dalam pengolahan data.
    

5. Setelah data diolah, saya menarik kesimpulan bahwa:
    - nasabah dengan jumlah anak yang lebih sedikit cenderung memiliki persentase gagal bayar yang lebih rendah, dan sebaliknya
    - semakin tinggi pendapatan nasabah, maka persentase gagal bayarnya cenderung semakin rendah, dan sebaliknya
    - nasabah yang sudah pernah menikah (married, divorced, dan widow / widower) memiliki memiliki persentase gagal bayar yang lebih kecil dibandingkan yang belum pernah menikah (civil partnership dan unmarried), dan sebaliknya
    - nasabah dengan tujuan membelanjakan uangnya ke bidang bangunan (rumah, properti, atau realestate) memiliki persentase gagal bayar yang lebih kecil dibandingkan tujuan lain
    - nasabah dengan tujuan car dan education memiliki kecenderungan gagal bayar yang tinggi

Berdasarkan kesimpulan tersebut dapat diberikan solusi berupa diberlakukannya pengketatan seleksi penilaian kredit untuk nasabah dengan kriteria:
- nasabah dengan jumlah anak yang cukup banyak
- nasabah dengan pendapatan rendah
- nasabah yang belum pernah menikah (civil partnership dan unmarried)
- nasabah dengan tujuan car dan education

<div class="alert alert-success">
<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Project ini merupakan lompatan yang sangat jauh dibandingkan project sebelumnya maka dari itu saya berterima kasih telah memberikan waktu dan usaha yang tidak sedikit dalam pengerjaannya. Namun masih ada beberapa catatan yang perlu dilakukan dalam perbaikan project ini. 
</div>